Update Logging In section of Testing Your App page (#4885)
cypress-io/cypress-documentation: content/guides/end-to-end-testing/testing-your-app.md

Luckily, Cypress provides the [`cy.session()`](/api/commands/session) command, a
powerful performance tool that lets you cache the browser context associated
with your user and reuse it for multiple tests without going through multiple
login flows! Let's modify the custom `cy.login()` command from our previous
example to use `cy.session()`:

```js
Cypress.Commands.add('login', (username, password) => {
  cy.session(
    username,
    () => {
      cy.visit('/login')
      cy.get('input[name=username]').type(username)
      cy.get('input[name=password]').type(`${password}{enter}`, { log: false })
      cy.url().should('include', '/dashboard')
      cy.get('h1').should('contain', username)
    },
    {
      validate: () => {
        cy.getCookie('your-session-cookie').should('exist')
      },
    }
  )
})
```

<Alert type="info">

<strong class="alert-header">Third-Party Login</strong>

If your app implements login via a third-party authentication provider such as
[Auth0](https://auth0.com/) or [Okta](https://www.okta.com/), you can use the
[`cy.origin()`](/api/commands/origin) command to include their login pages as
part of your authentication tests.

> </Alert>

Reviewer: does it make sense to plug the guides here?

Author: Problem is they don't have their own section to link to - we should probably change that

> 

Reviewer 2: The `Step 2: Visit your server` section is out of date with how Cypress works. Prob out of scope for this PR but we need to get an issue logged to get it updated with how Cypress 10+ works.

Author: OK I've made a note for later

Each of these likely requires a full blown e2e test.

#### Reusing the login code

At this point there's nothing stopping you copying and pasting the login code
above into every one of your tests that needs an authenticated user. Or you
could even put all your tests in one big spec file and put the login code in a
`beforeEach` block. But neither of those approaches is particularly
maintainable, and they're certainly not very elegant. A much better solution is
to use a `cy.login()` [custom command](/api/cypress-api/custom-commands).

Custom commands allow you to easily encapsulate and reuse Cypress test logic.

> They're basically just functions, but with some added sugar to allow you to use

Reviewer 2: I don't love "basically just functions". It'd also be worth mentioning that custom commands are also queued to execute async at a later time https://docs.cypress.io/guides/core-concepts/introduction-to-cypress#The-Cypress-Command-Queue

them with the same chaining API as the built-in Cypress commands. Here's an
example of the login code from the previous section converted into a custom
command:

> 

Reviewer 2: ```suggestion
Lets make the above login example a custom command, we added to `cypress/support/commands.js` so it can be leveraged in any spec file:
```

```js
// In cypress/support/commands.js

Cypress.Commands.add('login', (username, password) => {
  cy.visit('/login')

  cy.get('input[name=username]').type(username)

  // {enter} causes the form to submit
  cy.get('input[name=password]').type(`${password}{enter}`)

  // we should be redirected to /dashboard
  cy.url().should('include', '/dashboard')

  // our auth cookie should be present
  cy.getCookie('your-session-cookie').should('exist')

  // UI should reflect this user being logged in
  cy.get('h1').should('contain', username)
})

// In your spec file

it('does something on a secured page', function () {
  const { username, password } = this.currentUser
  cy.login(username, password)

  // ...rest of test
})
```

#### Improving performance

You're probably wondering what happened to our advice about logging in "only
once". The custom command above will work just fine for testing your secured
pages, but if you have more than a handful of tests, logging in before every

> test is going to add a lot of running time to your suite.

Reviewer 2: ```suggestion
test is going to add a lot of run time to your suite.
```
or
```suggestion
test is going to increase the overall run time of your suite.
```

> Luckily, Cypress provides the [`cy.session()` command](/api/commands/session), a

Reviewer 2: ```suggestion
Luckily, Cypress provides the [`cy.session()`](/api/commands/session) command, a
```

> powerful performance tool that lets you cache user session data and reuse it for

Reviewer 2: ```suggestion
powerful performance tool that lets you cache the browser context associated to your user's and reuse it for
```
session data is pretty overloaded and means different things to different users. Being explicit and saying cookies, local storage and session storage might be better.

> multiple tests without going through multiple login flows! Let's modify the

Reviewer 2: ```suggestion
multiple tests without going through multiple login flows! Let's modify the custom
```

`cy.login()` command from our previous example to use `cy.session()`:

```js
Cypress.Commands.add('login', (username, password) => {
  cy.session(
    username,
    () => {

>       cy.visit('/login')

Author: I think we should avoid repeating the comments, it blows out the length of the examples. This is a tutorial so it's intended that they read all the way through I think.

      cy.get('input[name=username]').type(username)

>       cy.get('input[name=password]').type(`${password}{enter}`)

Reviewer 2: ```suggestion
      cy.get('input[name=password]').type(`${password}{enter}`, { log: false }) // use {log: false} to prevent your password from showing in the Command Log 
```

      cy.url().should('include', '/dashboard')
      cy.get('h1').should('contain', username)
    },
    {
      validate: () => {
        cy.getCookie('your-session-cookie').should('exist')
      },
    }
  )
})
```

> 

Reviewer 2: I actually think it'd be worth adding a plug on cy.origin in this snippet. A lot of people use 3rd party auth provider and/or their auth route lives on a different origin that wouldn't work with this example. It could simply be "hey we have this commands to help! check out our authentication guides for more details". @AtofStryker has done some great updates on those guides

There's a lot going on here that's out of the scope of this introduction, so for
a more in-depth explanation of how `cy.session()` works,

> [read the API docs](/api/commands/session).

Reviewer 2: ```suggestion
[read the commands docs](/api/commands/session).
```
Cypress API is different than Cypress commands since its sync helpers that execute immediately instead of being queue to run later like Cypress commands

Author: I usually refer to everything underneath the "API" top-level docs tab as "the API docs". I know that there is also (confusingly) something called the "Cypress API" which is different, but I'm not referring to that. I do see your point but I think it needs to be consistent, if we're going to change this it needs to be "command docs" everywhere

> There's a lot going on here that's out of the scope of this introduction, so for
> a more in-depth explanation of how `cy.session()` works,
> [read the API docs](/api/commands/session).

Reviewer: we could also rephrase for something like this thats more explicit?
```suggestion
There's a lot going on here that's out of the scope for this introduction.
 Please check out [cy.session()](/api/commands/session) for
a more in-depth explanation.
```
